### ISSUE TITLE
Model Training error: Check failed: (shape_[0]) >= (end) Slice end index out of range

### ISSUE THREAD

**Issue by dipanjanS** [apache/incubator-mxnet#1978]
Facing issue with validation data during model fit. Works fine with training data but specifying eval data parameter throws an error. Using numpy arrays for input data.

Following are the configurations I am using.

```
data = mx.symbol.Variable('data')
fc1  = mx.symbol.FullyConnected(data, name='fc1', num_hidden=40)
act1 = mx.symbol.Activation(fc1, name='tanh1', act_type="tanh")
fc2  = mx.symbol.FullyConnected(act1, name = 'fc2', num_hidden = 35)
act2 = mx.symbol.Activation(fc2, name='tanh2', act_type="tanh")
fc3  = mx.symbol.FullyConnected(act2, name='fc3', num_hidden=35)
act3 = mx.symbol.Activation(fc3, name='tanh3', act_type="tanh")
fc4  = mx.symbol.FullyConnected(act3, name='fc4', num_hidden=30)
act4 = mx.symbol.Activation(fc4, name='tanh4', act_type="tanh")
fc5  = mx.symbol.FullyConnected(act4, name='fc5', num_hidden=2)
out  = mx.symbol.SoftmaxOutput(fc5, name = 'softmax')

model = mx.model.FeedForward(symbol=out, num_epoch=100, momentum=0.9, wd=0.00001, ctx=mx.cpu())
```

Following works perfectly fine

```
model.fit(X=train_df_matrix, y=train_labels_tf)
```

Following throws the error mentioned (`Check failed: (shape_[0]) >= (end) Slice end index out of range`)

```
model.fit(X=train_df_matrix, y=train_labels_tf, eval_data=(validate_df_matrix, validate_labels_tf))
```

Shape of objects are as follows
![image](https://cloud.githubusercontent.com/assets/3448263/14876332/7d6abf5a-0d31-11e6-8feb-bbdf17fffabe.png)

Any help will be really appreciated!

**Comment by rldotai** [comment]
In the future, an exception trace would be helpful. 
Can you try setting `numpy_batch_size=32` in your call to `model.fit()`?

**Comment by rldotai** [comment]
This issue seems to have occurred before, c.f #1281 and #1743, but #612 and #1093 seem to have solved it by adjusting their data iterator shapes.

Under the hood, MXNet creates two data iterators for the training set and the validation set.
What makes this really weird is that the code that creates both is more or less the same, once it does some type checking/conversion.
They should eventually both reach the same call in `mxnet.model.FeedForward._init_iter`:

```python
class FeedForward:
    ...
    def _init_iter(self, X, y, is_train):
        ....
        return io.NDArrayIter(X, y, min(X.shape[0], self.numpy_batch_size),
                                      shuffle=is_train, last_batch_handle='roll_over')
```

The default value for `numpy_batch_size` is 128, yet your testing dataset is smaller than that, so perhaps that's causing some sort of shape-mismatch?

`DataIterator` objects are kinda tedious to work with, especially if you're only doing it indirectly. MXNet really wants you to just make all of the inputs into `DataIterator` objects, and so far I've found everything runs more smoothly when I define the data's shape and the batch size explicitly.

**Comment by dipanjanS** [comment]
@rldotai Thanks for the detailed explanation, I will try looking into this, right now I am directly using numpy arrays not the `DataIterator` objects but internally also it does the transformation before model fitting even if I use numpy arrays is what I am gathering from your explanation?

I will also try setting the `numpy_batch_size` parameter and try it out. Also would you have a simple example or tutorial of how to convert numpy arrays into `DataIterator` objects? I am really having difficulty in that aspect hence I use numpy arrays directly since I am already using them for other ML related operations.

**Comment by rldotai** [comment]
In response to the first part: that would seem to be the case. As I understand things, using `DataIterator` for everything is a design decision to increase the overall speed and facilitate distributed training.

As for an example, how about MNIST with a convolutional net? I believe this is from one of the examples but I don't have the link handy right now.

```python
from sklearn.datasets import fetch_mldata
mnist = fetch_mldata('MNIST original')

# Convert to float and scale between [0,1]
X = X.astype(np.float32)/255

# Reshape into array of  1x28x28 arrays... the 1 corresponds to only gray color channel
X = X.reshape(-1, 1, 28, 28)

# Split the data into training/validation sets
X_train = X[:60000]
X_test = X[60000:]
Y_train = Y[:60000]
Y_test = Y[60000:]

# Create the iterators for the train/test arrays, note same batch size
batch_size = 128
train_iter = mx.io.NDArrayIter(X_train, Y_train, batch_size=batch_size)
test_iter = mx.io.NDArrayIter(X_test, Y_test, batch_size=batch_size)
```

We can check that the array looks right, so I didn't make a mistake w/ resizing

```python
>>> X.shape 
(70000, 1, 28, 28)
```

But we can also check what the batches look like:


```python
# Get a batch
>>> b = train_iter.next()
# Batches keep track of both data and labels, we want data here
>>> a = b.data[0]
>>> a.shape
(128, 1, 28, 28)
```

So you can relatively easily define the iterators yourself, and it gives you a bit more control over the training process.
Alternatively, you could try other frameworks like Theano/Keras or TensorFlow, or use those to prototype your network before using MXNet when speed/distributed training becomes more important.

**Comment by phunterlau** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
